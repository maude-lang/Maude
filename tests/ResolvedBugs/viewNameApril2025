#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/viewNameApril2025.maude -no-banner -no-advise \
  > viewNameApril2025.out 2>&1

diff $srcdir/viewNameApril2025.expected viewNameApril2025.out > /dev/null 2>&1
