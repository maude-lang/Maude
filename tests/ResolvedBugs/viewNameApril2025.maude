***
***	We had a bug because view names were unrestricted single token
***	identifers but they ended up inside instantiated sort names and
***	sort names have many restrictions. This is resolved by no longer
***	allowing ( ) [ ] { } , in view names and allowing any non-space
***	characters except these 7 inside of balanced {}s within sort names.
***

*** This should be rejected
view B`{`}AD from TRIV to NAT is
  sort Elt to Nat .
endv

fmod TEST is
  inc SET{B`{`}AD} .
endfm

*** Test : ending up inside sort name as a result of instantiation
view NAT::AS::TRIV from TRIV to NAT is
  sort Elt to Nat .
endv

fmod TEST2 is
  inc SET{NAT::AS::TRIV} .
  var X : NeSet{NAT::AS::TRIV} .
  var Y : NeSet`{NAT::AS::TRIV`} .  *** single token version
endfm

show sorts .

red X2:NeSet{NAT::AS::TRIV} == (X).NeSet{NAT::AS::TRIV} .
red Y2:NeSet`{NAT::AS::TRIV`} == (Y).NeSet`{NAT::AS::TRIV`} .

*** Test . ending up inside sort name as a result of instantiation
view NAT.AS.TRIV from TRIV to NAT is
  sort Elt to Nat .
endv

fmod TEST3 is
  inc SET{NAT.AS.TRIV} .
  var X : NeSet{NAT.AS.TRIV} .
  var Y : NeSet`{NAT.AS.TRIV`} .  *** single token version
endfm

show sorts .

red X2:NeSet{NAT.AS.TRIV} == (X).NeSet{NAT.AS.TRIV} .
red Y2:NeSet`{NAT.AS.TRIV`} == (Y).NeSet`{NAT.AS.TRIV`} .

*** Bad sort names
fmod TEST4{X :: TRIV} is
  sorts `[Bad`] B`{ad B`,a`,d`{X`} B:a:d B.a.d{X} .
endfm

show sorts .

*** Make sure it can't be used at the metalevel
red in META-LEVEL : metaReduce(['TEST4], 'true.Bool) .

*** Bad implied sort name
fmod TEST5 is
  op f : `[Bad`] -> `[Bad`] .
endfm

*** Bad class name
omod TEST6 is
  class `[Bad`] .
endom

*** OK sort names in single token form
fmod TEST7{X :: TRIV} is
  sorts Ok`{X:Y`} Ok`{X.Y`} Ok`{X`,Y`}`{X.Y`} .
endfm

show sorts .

*** OK sort names in pretty form
fmod TEST8{X :: TRIV} is
  sorts Ok{X:Y} Ok{X.Y} Ok{X,Y}{X.Y} .
endfm

show sorts .
