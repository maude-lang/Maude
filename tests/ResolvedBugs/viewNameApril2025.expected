Warning: <standard input>, line 12 (view B`{`}AD): B`{`}AD is not a valid view
    name.
Warning: <standard input>, line 17 (fmod TEST): unusable view B`{`}AD.
Warning: <standard input>, line 16 (fmod TEST): this module contains one or
    more errors that could not be patched up and thus it cannot be used or
    imported.
sort Bool .
sort Zero .	subsorts Zero < Nat NeSet{NAT::AS::TRIV} Set{NAT::AS::TRIV} .
sort NzNat .	subsorts NzNat < Nat NeSet{NAT::AS::TRIV} Set{NAT::AS::TRIV} .
sort Nat .	subsorts NzNat Zero < Nat < NeSet{NAT::AS::TRIV} Set{
    NAT::AS::TRIV} .
sort NeSet{NAT::AS::TRIV} .	subsorts NzNat Zero Nat < NeSet{NAT::AS::TRIV}
    < Set{NAT::AS::TRIV} .
sort Set{NAT::AS::TRIV} .	subsorts NzNat Zero Nat NeSet{NAT::AS::TRIV} <
    Set{NAT::AS::TRIV} .
==========================================
reduce in TEST2 : X2:NeSet{NAT::AS::TRIV} == X .
rewrites: 1
result Bool: false
==========================================
reduce in TEST2 : Y2:NeSet{NAT::AS::TRIV} == Y .
rewrites: 1
result Bool: false
sort Bool .
sort Zero .	subsorts Zero < Nat NeSet{NAT.AS.TRIV} Set{NAT.AS.TRIV} .
sort NzNat .	subsorts NzNat < Nat NeSet{NAT.AS.TRIV} Set{NAT.AS.TRIV} .
sort Nat .	subsorts NzNat Zero < Nat < NeSet{NAT.AS.TRIV} Set{NAT.AS.TRIV}
    .
sort NeSet{NAT.AS.TRIV} .	subsorts NzNat Zero Nat < NeSet{NAT.AS.TRIV} <
    Set{NAT.AS.TRIV} .
sort Set{NAT.AS.TRIV} .	subsorts NzNat Zero Nat NeSet{NAT.AS.TRIV} < Set{
    NAT.AS.TRIV} .
==========================================
reduce in TEST3 : X2:NeSet{NAT.AS.TRIV} == X .
rewrites: 1
result Bool: false
==========================================
reduce in TEST3 : Y2:NeSet{NAT.AS.TRIV} == Y .
rewrites: 1
result Bool: false
Warning: <standard input>, line 54 (fmod TEST4): `[Bad`] is not a valid sort
    name.
Warning: <standard input>, line 54 (fmod TEST4): B`{ad is not a valid sort
    name.
Warning: <standard input>, line 54 (fmod TEST4): B`,a`,d`{X`} is not a valid
    sort name.
Warning: <standard input>, line 54 (fmod TEST4): B:a:d is not a valid sort
    name.
Warning: <standard input>, line 54 (fmod TEST4): B.a.d`{X`} is not a valid sort
    name.
Warning: <standard input>, line 53 (fmod TEST4): this module contains one or
    more errors that could not be patched up and thus it cannot be used or
    imported.
Warning: <standard input>, line 53 (fmod TEST4): module TEST4 is unusable due
    to unpatchable errors.
==========================================
reduce in META-LEVEL : metaReduce(['TEST4], 'true.Bool) .
Warning: <metalevel>: unable to use module TEST4 due to unpatchable errors.
rewrites: 1
result [ResultPair?]: metaReduce(sth 'TEST4 is
  including 'TEST4 .
  sorts none .
  none
  none
  none
  none
  none
  none
  none
endsth, 'true.Bool)
Warning: <standard input>, line 64 (fmod TEST5): `[Bad`] is not a valid sort
    name.
Warning: <standard input>, line 64 (fmod TEST5): undeclared sort `[Bad`].
Warning: <standard input>, line 63 (fmod TEST5): this module contains one or
    more errors that could not be patched up and thus it cannot be used or
    imported.
Warning: <standard input>, line 69 (omod TEST6): `[Bad`] is not a valid class
    name.
Warning: <standard input>, line 68 (omod TEST6): this module contains one or
    more errors that could not be patched up and thus it cannot be used or
    imported.
sort X$Elt .
sort Bool .
sort Ok{X:Y} .
sort Ok{X.Y} .
sort Ok{X,Y}{X.Y} .
sort X$Elt .
sort Bool .
sort Ok{X:Y} .
sort Ok{X.Y} .
sort Ok{X,Y}{X.Y} .
Bye.
